Safe Access to Contiguous Storage (#2307)
apple/swift-evolution: proposals/nnnn-safe-shared-contiguous-storage.md

extension Array {
  // note: this could borrow a temporary copy of the `Array`'s storage
  var storage: Span<Element> { _read }
}
extension ArraySlice {
  // note: this could borrow a temporary copy of the `ArraySlice`'s storage
  var storage: Span<Element> { _read }
}
extension ContiguousArray {
  var storage: Span<Element> { get }
}

extension Foundation.Data {
  var storage: Span<UInt8> { get }
}

extension String.UTF8View {
  // note: this could borrow a temporary copy of the `String`'s storage
  var storage: Span<Unicode.UTF8.CodeUnit> { _read }
}
extension Substring.UTF8View {
  // note: this could borrow a temporary copy of the `Substring`'s storage
  var storage: Span<Unicode.UTF8.CodeUnit> { _read }
}
extension Character.UTF8View {
  // note: this could borrow a temporary copy of the `Character`'s storage
  var storage: Span<Unicode.UTF8.CodeUnit> { _read }
}

extension SIMD {
  var storage: Span<Self.Scalar> { get }
}
extension KeyValuePairs {
  var storage: Span<(Self.Key, Self.Value)> { get }
}
extension CollectionOfOne {
  var storage: Span<Element> { get }
}

extension Slice: ContiguousStorage where Base: ContiguousStorage {
  var storage: Span<Base.Element> { get }

> }

Reviewer: The `ContiguousStorage` protocol has been moved to "Future directions", but it's still mentioned in the "Detailed design".

Author: Removed this.

  
  public func extracting(_ bounds: some RangeExpression<Int>) -> Self
  
  public func extracting(
    uncheckedBounds bounds: some RangeExpression<Int>
  ) -> Self

  public func extracting(_: UnboundedRange) -> Self

  // extracting prefixes and suffixes

  public func extracting(first maxLength: Int) -> Self

  public func extracting(droppingLast k: Int) -> Self

  public func extracting(last maxLength: Int) -> Self

  public func extracting(droppingFirst k: Int = 1) -> Self
}
```

## Source compatibility

This proposal is additive and source-compatible with existing code.

## ABI compatibility

This proposal is additive and ABI-compatible with existing code.

## Implications on adoption

The additions described in this proposal require a new version of the standard library and runtime.

## <a name="Alternatives"></a>Alternatives considered

##### Make `Span` a noncopyable type
Making `Span` non-copyable was in the early vision of this type. However, we found that would make `Span` a poor match to model borrowing semantics. This realization led to the initial design for non-escapable declarations.

##### A protocol in addition to `ContiguousStorage` for unsafe buffers

> This document proposes adding the `ContiguousStorage` protocol to the standard library's `Unsafe{Mutable,Raw}BufferPointer` types. On the surface this seems like whitewashing the unsafety of these types. The lifetime constraint only applies to the binding used to obtain a `Span`, and the initialization precondition can only be enforced by documentation. Nothing will prevent unsafe code from deinitializing a portion of the storage while a `Span` is alive. There is no safe bridge from `UnsafeBufferPointer` to `ContiguousStorage`. We considered having the unsafe buffer types conforming to a different version of `ContiguousStorage`, which would vend a `Span` through a closure-taking API. Unfortunately such a closure would be perfectly capable of capturing the `UnsafeBufferPointer` binding and be as unsafe as can be. For this reason, the `UnsafeBufferPointer` family will conform to `ContiguousStorage`, with safety being enforced in documentation.

Reviewer: The `ContiguousStorage` protocol has been moved to "Future directions", but it's still mentioned in "Alternatives considered".

Author: I am removing this as well, replacing it with a short note at the end of the `ContiguousStorage` protocol future direction. Thanks for pointing out the mistake @benrimmington!

Loading an uninitialized non-`BitwiseCopyable` value leads to undefined behavior. Loading an uninitialized `BitwiseCopyable` value does not immediately lead to undefined behavior, but it produces a garbage value which may lead to misbehavior of the program.

A `MutableSpan<T>` should provide a better, safer alternative to mutable memory in the same way that `Span<T>` provides a better, safer read-only type. `MutableSpan<T>` would also automatically enforce exclusivity of writes.

However, it alone does not track initialization state of each address, and that will continue to be the responsibility of the developer.

##### Delegating initialization of memory with `OutputSpan<T>`

Some data structures can delegate initialization of their initial memory representation, and in some cases the initialization of additional memory. In the standard library we have `Array.init(unsafeUninitializedCapacity:initializingWith:)` and `String.init(unsafeUninitializedCapacity:initializingUTF8With:)`. A safer abstraction for initialization would make such initializers less dangerous, and would allow for a greater variety of them.

> `OutputSpan<T>` would need run-time bookkeeping (e.g. a bitvector with a bit per-address) to track initialization state to safely support random access and random-order initialization.

Reviewer 2: I very strongly object to any bitmap in `OutputSpan`. `OutputSpan` needs to work like a fixed-capacity array type over someone else's storage -- initialized items must all be placed in a contiguous prefix region.

Author: The first-line `OutputSpan` won't have a bitmap. A version with a bitmap would be for targeted cases when initializing in order isn't practical.

Author: I updated this section to actually describe the initial `OutputSpan<T>`.

> Alternatively, a divide-and-conqueor style initialization order might be solvable via an API layer without run-time bookkeeping, but with more complex ergonomics.

Reviewer 2: The primary use case of `OutputSpan` is to serve as the unit of data exchange in consuming iteration. (E.g., to allow containers of noncopyable types to efficiently implement an `append` operation that takes a generic consuming sequence.)

Allowing divide-and-conquer initialization is a second order benefit, at best. Luckily, that doesn't need bitmaps, either.

Author: improved the description of `OutputSpan`.

##### <a name="Bytes"></a>Resizable, contiguously-stored, untyped collection in the standard library

The example in the [motivation](#motivation) section mentions the `Foundation.Data` type. There has been some discussion of either replacing `Data` or moving it to the standard library. This document proposes neither of those. A major issue is that in the "traditional" form of `Foundation.Data`, namely `NSData` from Objective-C, it was easier to control accidental copies because the semantics of the language did not lead to implicit copying. 

Even if `Span` were to replace all uses of a constant `Data` in API, something like `Data` would still be needed, just as `Array<T>` will: resizing mutations (e.g. `RangeReplaceableCollection` conformance.) We may still want to add an untyped-element equivalent of `Array` at a later time.

##### <a name="Conversions"></a>Syntactic Sugar for Automatic Conversions

> In the context of a resilient library, a generic entry point in terms of `some ContiguousStorage` may add unwanted overhead. As detailed above, an entry point in an evolution-enabled library requires an inlinable  generic public entry point which forwards to a publicly-accessible function defined in terms of `Span`. If `Span` does become a widely-used type to interface between libraries, we could simplify these conversions with a bit of compiler help.

Reviewer 2: I am still wary of encouraging folks to define API that takes single spans, as opposed to arbitrary borrowing sequences.

Author: This one is getting more and more speculative, since even `ContiguousStorage` has to be pushed back!

  return b.distance(from: b.startIndex, to: e)
}
let data = Data((0..<9).reversed()) // Data conforms to ContiguousStorage
let array = Array(data) // Array<UInt8> also conforms to ContiguousStorage
myStrnlen(data)  // 8
myStrnlen(array) // 8
```

This would probably consist of a new type of custom conversion in the language. A type author would provide a way to convert from their type to an owned `Span`, and the compiler would insert that conversion where needed. This would enhance readability and reduce boilerplate.

##### Interopability with C++'s `std::span` and with llvm's `-fbounds-safety`
The [`std::span`](https://en.cppreference.com/w/cpp/container/span) class template from the C++ standard library is a similar representation of a contiguous range of memory. LLVM may soon have a [bounds-checking mode](https://discourse.llvm.org/t/70854) for C. These are opportunities for better, safer interoperation with Swift, via a type such as `Span`.

## Acknowledgments

Joe Groff, John McCall, Tim Kientzle, Karoy Lorentey contributed to this proposal with their clarifying questions and discussions.

## <a name="Indexing"></a>Appendix: Index and slicing design considerations

> Early prototypes of this proposal defined an `Index` type, `Iterator` types, etc. We are proposing `Int`-based API and are deferring defining `Index` and `Iterator` until more of the non-escapable collection story is sorted out. The below is some of our research into different potential designs of an `Index` type.

Reviewer 2: We are okay with having an `Index` type on `Span`, and we must be proposing it above. (`UnsafeBufferPointer` has sunk the concrete concept of an `Index` down into the ownership model in SE-0437. Whatever shape we will propose for a collection-like container of noncopyable elements, it will include the idea of an escaping `Index`, with this exact name.)

(`Iterator` on the other hand is not the right name for a borrowing iterator type, as the classic `Sequence` protocol reserves this name for its unified borrowing+consuming iterator. That iterator shape does not translate to the ownership model. `Sequence` will be split into two+ variants, with each variant coming with a distinct name for its iterators. `Span` will soon need to define a borrowing iterator type; however, this work must be deferred until we have introduced that concept -- the addition should be done in that same proposal.)

There are 3 potentially-desirable features of `Span`'s `Index` design:

1. `Span` is its own slice type
2. Indices from a slice can be used on the base collection
3. Additional reuse-after-free checking

Each of these introduces practical tradeoffs in the design.

> #### `Span` is its own slice type

Reviewer 2: This section should be removed. There is no possible future direction where `Span` becomes its own slice type. (Or, alternatively, we are already living in that world by having migrated to `extracting`.)

The concept of a slice is tied to the `Collection` protocol, and it does not appear to survive the removal of the assumption of copyability -- at least not without radical changes/restrictions.

* Review Manager: TBD
* Status: **Awaiting implementation**
* Roadmap: [BufferView Roadmap](https://forums.swift.org/t/66211)
* Bug: rdar://48132971, rdar://96837923
* Implementation: Prototyped in https://github.com/apple/swift-collections (on branch "future")
* Review: ([pitch 1](https://forums.swift.org/t/69888))([pitch 2](https://forums.swift.org/t/72745))

## Introduction

We introduce `Span<T>`, an abstraction for container-agnostic access to contiguous memory. It will expand the expressivity of performant Swift code without compromising on the memory safety properties we rely on: temporal safety, spatial safety, definite initialization and type safety.

In the C family of programming languages, memory can be shared with any function by using a pointer and (ideally) a length. This allows contiguous memory to be shared with a function that doesn't know the layout of a container being used by the caller. A heap-allocated array, contiguously-stored named fields or even a single stack-allocated instance can all be accessed through a C pointer. We aim to enable a similar idiom in Swift, without compromising Swift's memory safety.

> This proposal is related to two compiler features: [Nonescapable types](https://github.com/swiftlang/swift-evolution/pull/2304) (`~Escapable`,) being proposed along side, and [Compile-time Lifetime Dependency Annotations][PR-2305], which will be proposed in the following weeks. A precursor to this proposal was the [BufferView roadmap](https://forums.swift.org/t/66211) forum thread. This proposal is also related to the following proposals:

Author: I've adjusted this – thanks.

```swift
extension Span where Element: ~Copyable & ~Escapable {
  public var count: Int { get }
  public var isEmpty: Bool { get }

  subscript(_ position: Int) -> Element { _read }
}
```

Note that `Span` does _not_ conform to `Collection`. This is because `Collection`, as originally conceived and enshrined in existing source code, assumes pervasive copyability and escapability of the `Collection` itself as well as of element type. In particular a subsequence of a `Collection` is semantically a separate value from the instance it was derived from. In the case of `Span`, a sub-span representing a subrange of its elements _must_ have the same lifetime as the `Span` from which it originates. Another proposal will consider collection-like protocols to accommodate different combinations of `~Copyable` and `~Escapable` for the collection and its elements.

`Span`s representing subsets of consecutive elements can be extracted out of a larger `Span` with an API similar to the `extracting()` functions recently added to `UnsafeBufferPointer` in support of non-copyable elements:

```swift
extension Span where Element: ~Copyable & ~Escapable {

>   public func extracting(_ bounds: Range<Int>) -> Self

Reviewer 3: I think it will work against us to include any APIs whose semantics require talking about lifetime dependencies. It will open of a line of discussion that we can't resolve and will cause delays. If we're going to split Span into two pitches, then let's restrict the lifetime dependence discussion to the second pitch and focus on that second pitch as soon as possible.

>   public func extracting(_ bounds: Range<Int>) -> Self

Author: These are moved to a future directions item.

```swift
extension UnsafeBufferPointer { // as well as UnsafeMutableBufferPointer
  public func withUnsafeSpan<E: Error, Result: ~Copyable>(
    _ body: (_ elements: Span<Element>) throws(E) -> Result
  ) throws(E) -> Result

  public func withUnsafeBytes<E: Error, Result: ~Copyable>(
    _ body: (_ elements: RawSpan) throws(E) -> Result
  ) throws(E) -> Result where Element: BitwiseCopyable
}

extension UnsafeRawBufferPointer { // as well as UnsafeMutableRawBufferPointer
  public func withUnsafeBytes<E: Error, Result: ~Copyable>(
    _ body: (_ elements: RawSpan) throws(E) -> Result
  ) throws(E) -> Result
}
```

`Span` and `RawSpan` also have an unsafe hatch for use with functions that require an unsafe argument, such as C functions:

```swift
extension Span where Element: ~Copyable & ~Escapable {

>   func withUnsafeBufferPointer<E: Error, Result: ~Copyable & ~Escapable>(

Reviewer 3: `Span.withUnsafeBufferPointer` ensures that the Span's lifetime, and whatever that span depends on, extend across all *valid* uses of the pointer. The rules for pointer validity in closure taking methods are clear. There's nothing new here. As long as programmers follow the usual rules, they can't introduce any new form of unsafety.
The compiler should be able to diagnose pointer escapes from the closure, but that's a separate discussion!

  func withUnsafeBytes<E: Error, Result: ~Copyable & ~Escapable>(
    _ body: (_ buffer: UnsafeRawBufferPointer) throws(E) -> Result
  ) throws(E) -> Result
}

extension RawSpan {
  func withUnsafeBytes<E: Error, Result: ~Copyable & ~Escapable>(
    _ body: (_ buffer: UnsafeRawBufferPointer) throws(E) -> Result
  ) throws(E) -> Result
}
```

### Complete `Span` API:

```swift
public struct Span<Element: ~Copyable & ~Escapable>: Copyable, ~Escapable {
  internal var _start: UnsafePointer<Element>
  internal var _count: Int
}
```

##### Basic API:

The following properties, functions and subscripts have direct counterparts in the `Collection` protocol hierarchy. Their semantics shall be as described where they counterpart is declared (in `Collection` or `RandomAccessCollection`). The only difference with their counterpart should be a lifetime dependency annotation where applicable, allowing them to return borrowed nonescapable values or borrowed noncopyable values.

```swift
extension Span where Element: ~Copyable & ~Escapable {
  public var count: Int { get }
  public var isEmpty: Bool { get }

  public subscript(_ position: Int) -> Element { _read }
}
```

> ##### Accessing subranges of elements:

Reviewer 4: As I noted earlier on, all of the `extracting` functions depend on lifetime dependencies. Perhaps provide a closure-based `withSubspan`?

# Safe Access to Contiguous Storage

* Proposal: [SE-NNNN](nnnn-safe-shared-contiguous-storage.md)
* Authors: [Guillaume Lessard](https://github.com/glessard), [Andrew Trick](https://github.com/atrick), [Michael Ilseman](https://github.com/milseman)
* Review Manager: TBD
* Status: **Awaiting implementation**
* Roadmap: [BufferView Roadmap](https://forums.swift.org/t/66211)
* Bug: rdar://48132971, rdar://96837923
* Implementation: Prototyped in https://github.com/apple/swift-collections (on branch "future")
* Review: ([pitch 1](https://forums.swift.org/t/69888))([pitch 2](https://forums.swift.org/t/72745))

## Introduction

> We introduce `Span<T>`, an abstraction for container-agnostic access to contiguous memory. It will expand the expressivity of performant Swift code without comprimising on the memory safety properties we rely on: temporal safety, spatial safety, definite initialization and type safety.

Reviewer: ```suggestion
We introduce `Span<T>`, an abstraction for container-agnostic access to contiguous memory. It will expand the expressivity of performant Swift code without compromising on the memory safety properties we rely on: temporal safety, spatial safety, definite initialization and type safety.
```

In the C family of programming languages, memory can be shared with any function by using a pointer and (ideally) a length. This allows contiguous memory to be shared with a function that doesn't know the layout of a container being used by the caller. A heap-allocated array, contiguously-stored named fields or even a single stack-allocated instance can all be accessed through a C pointer. We aim to create a similar idiom in Swift, without compromising Swift's memory safety.

> This proposal is related to two other features being proposed along with it: [Nonescapable types](https://github.com/apple/swift-evolution/pull/2304) (`~Escapable`) and [Compile-time Lifetime Dependency Annotations](https://github.com/swiftlang/swift-evolution/pull/2305), as well as related to the [BufferView roadmap](https://forums.swift.org/t/66211) forum thread. This proposal is also related to the following proposals:

Reviewer: ```suggestion
This proposal is related to two other features being proposed along with it: [Nonescapable types](https://github.com/swiftlang/swift-evolution/pull/2304) (`~Escapable`) and [Compile-time Lifetime Dependency Annotations](https://github.com/swiftlang/swift-evolution/pull/2305), as well as related to the [BufferView roadmap](https://forums.swift.org/t/66211) forum thread. This proposal is also related to the following proposals:
```

By relying on borrowing, `Span` can provide simultaneous access to a non-copyable container, and can help avoid unwanted copies of copyable containers. Note that `Span` is not a replacement for a copyable container with owned storage; see the future directions for more details ([Resizable, contiguously-stored, untyped collection in the standard library](#Bytes))

`Span` is the currency type for local processing over values in contiguous memory. It is the replacement for any API currently using `Array`, `UnsafeBufferPointer`, `Foundation.Data`, etc., that does not need to escape the value.

### `RawSpan`

> `RawSpan` allows sharing the contiguous internal representation for values which may be heterogenously-typed, such as in decoders. Since it is a fully concrete type, it can achieve better performance in debug builds of client code as well as a more straightforward understanding of performance in library code.

Reviewer: ```suggestion
`RawSpan` allows sharing the contiguous internal representation for values which may be heterogeneously-typed, such as in decoders. Since it is a fully concrete type, it can achieve better performance in debug builds of client code as well as a more straightforward understanding of performance in library code.
```

  public var count: Int { get }
  public var isEmpty: Bool { get }

  subscript(_ position: Int) -> Element { _read }
}
```

Note that `Span` does _not_ conform to `Collection`. This is because `Collection`, as originally conceived and enshrined in existing source code, assumes pervasive copyability and escapability for itself as well as its elements. In particular a subsequence of a `Collection` is semantically a separate value from the instance it was derived from. In the case of `Span`, a sub-span representing a subrange of its elements _must_ have the same lifetime as the `Span` from which it originates. Another proposal will consider collection-like protocols to accommodate different combinations of `~Copyable` and `~Escapable` for the collection and its elements.

`Span`s representing subsets of consecutive elements can be extracted out of a larger `Span` with an API similar to the recently added `extracting()` functions of `UnsafeBufferPointer`:

```swift
extension Span where Element: ~Copyable & ~Escapable {
  public func extracting(_ bounds: Range<Int>) -> Self
}
```

The first element of a given span is _always_ at position zero, and its last it as position `count-1`.

> As a side-effect of not conforming to `Collection` or `Sequence`, `Span` is not directly supported by `for` loops at this time. It is, however, easy to use in a `for` loop via indexing:

Reviewer 5: Why does it not conform to `Collection` or `Sequence`? (I assume it's a non-copyable types issue, but it'd be good to explain it here.)

Reviewer 6: This is explained in the introduction:
> Note that Span does not conform to Collection. This is because Collection, as originally conceived and enshrined in existing source code, assumes pervasive copyability and escapability for itself as well as its elements. In particular a subsequence of a Collection is semantically a separate value from the instance it was derived from. In the case of Span, a sub-span representing a subrange of its elements must have the same lifetime as the Span from which it originates. Another proposal will consider collection-like protocols to accommodate different combinations of ~Copyable and ~Escapable for the collection and its elements.

  /// - Returns: A span with at most `maxLength` elements.
  public func extracting(last maxLength: Int) -> Self

  /// Returns a span over all but the given number of initial elements.
  ///
  /// If the number of elements to drop exceeds the number of elements in
  /// the span, the result is an empty span.
  ///
  /// The returned span's first item is always at offset 0; unlike buffer
  /// slices, extracted spans do not share their indices with the
  /// span from which they are extracted.
  ///
  /// - Parameter k: The number of elements to drop from the beginning of
  ///   the span. `k` must be greater than or equal to zero.
  /// - Returns: A span starting after the specified number of elements.
  public func extracting(droppingFirst k: Int) -> Self
}
```

##### Identifying whether a span is a subrange of another:

When working with multiple `Span` instances, it is often desirable to know whether one is a subrange of another. We include a function to determine whether this is the case, as well as a function to obtain the valid offsets of the subrange within the larger span:

```swift
extension Span where Element: ~Copyable & ~Escapable {

> 	/// Returns true if the memory represented by `span` is a subrange of

Reviewer: ```suggestion
  /// Returns true if the memory represented by `span` is a subrange of
```

  /// the memory represented by `self`
  ///
  /// Parameters:
  /// - span: a span of the same type as `self`
  /// Returns: whether `span` is a subrange of `self`
  public func contains(_ span: borrowing Self) -> Bool
  

> 	/// Returns the offsets where the memory of `span` is located within

Reviewer: ```suggestion
  /// Returns the offsets where the memory of `span` is located within
```

  /// - Returns: The return value of the `body` closure parameter.
  func withUnsafeBufferPointer<E: Error, Result: ~Copyable>(
    _ body: (_ buffer: UnsafeBufferPointer<Element>) throws(E) -> Result
  ) throws(E) -> Result
}

extension Span where Element: BitwiseCopyable {
  /// Calls the given closure with a pointer to the underlying bytes of
  /// the viewed contiguous storage.
  ///
  /// The buffer pointer passed as an argument to `body` is valid only
  /// during the execution of `withUnsafeBytes(_:)`.
  /// Do not store or return the pointer for later use.
  ///
  /// - Parameter body: A closure with an `UnsafeRawBufferPointer`
  ///   parameter that points to the viewed contiguous storage.
  ///   If `body` has a return value, that value is also
  ///   used as the return value for the `withUnsafeBytes(_:)` method.
  ///   The closure's parameter is valid only for the duration of
  ///   its execution.
  /// - Returns: The return value of the `body` closure parameter.
  func withUnsafeBytes<E: Error, Result: ~Copyable>(
    _ body: (_ buffer: UnsafeRawBufferPointer) throws(E) -> Result
  ) throws(E) -> Result
}
```

### RawSpan

> In addition to `Span<T>`, we propose the addition of `RawSpan`, to represent heterogenously-typed values in contiguous memory. `RawSpan` is similar to `Span<T>`, but represents _untyped_ initialized bytes. `RawSpan` is a specialized type that intends to support parsing and decoding applications, as well as applications where heavily-used code paths require concrete types as much as possible. Its API supports extracting sub-spans, along with the operations `load(as:)` and `loadUnaligned(as:)`.

Reviewer: ```suggestion
In addition to `Span<T>`, we propose the addition of `RawSpan`, to represent heterogeneously-typed values in contiguous memory. `RawSpan` is similar to `Span<T>`, but represents _untyped_ initialized bytes. `RawSpan` is a specialized type that intends to support parsing and decoding applications, as well as applications where heavily-used code paths require concrete types as much as possible. Its API supports extracting sub-spans, along with the operations `load(as:)` and `loadUnaligned(as:)`.
```

  ///     with the value in the memory referenced by this pointer.
  public func unsafeLoad<T>(
    fromUncheckedByteOffset offset: Int, as: T.Type
  ) -> T

  /// Returns a new instance of the given type, constructed from the raw memory
  /// at the specified offset.
  ///
  /// The memory at this pointer plus `offset` must be initialized to `T`
  /// or another type that is layout compatible with `T`.
  ///
  /// This is an unsafe operation. This function does not validate the bounds
  /// of the memory access, and failure to meet the preconditions
  /// above may produce an invalid value of `T`.
  ///
  /// - Parameters:
  ///   - offset: The offset from this pointer, in bytes. `offset` must be
  ///     nonnegative. The default is zero.
  ///   - type: The type of the instance to create.
  /// - Returns: A new instance of type `T`, read from the raw bytes at
  ///     `offset`. The returned instance isn't associated
  ///     with the value in the range of memory referenced by this pointer.
  public func unsafeLoadUnaligned<T: BitwiseCopyable>(
    fromUncheckedByteOffset offset: Int, as: T.Type
  ) -> T
}
```

> A `RawSpan` can be viewed as a `Span<T>`, provided the memory is laid out homogenously as instances of `T`.

Reviewer: ```suggestion
A `RawSpan` can be viewed as a `Span<T>`, provided the memory is laid out homogeneously as instances of `T`.
```

We could provide an automatic way to use a `ContiguousStorage`-conforming type with a function that takes a `Span` of the appropriate element type:

```swift
func myStrnlen(_ b: Span<UInt8>) -> Int {
  guard let i = b.firstIndex(of: 0) else { return b.count }
  return b.distance(from: b.startIndex, to: e)
}
let data = Data((0..<9).reversed()) // Data conforms to ContiguousStorage
let array = Array(data) // Array<UInt8> also conforms to ContiguousStorage
myStrnlen(data)  // 8
myStrnlen(array) // 8
```

This would probably consist of a new type of custom conversion in the language. A type author would provide a way to convert from their type to an owned `Span`, and the compiler would insert that conversion where needed. This would enhance readability and reduce boilerplate.

> #### Interopability with C++'s `std::span` and with llvm's `-fbounds-safety`

Reviewer: ```suggestion
#### Interoperability with C++'s `std::span` and with llvm's `-fbounds-safety`
```

## Acknowledgments

Joe Groff, John McCall, Tim Kientzle, Karoy Lorentey contributed to this proposal with their clarifying questions and discussions.

### <a name="Indexing"></a>Appendix: Index and slicing design considerations

Early prototypes of this proposal defined an `Index` type, `Iterator` types, etc. We are proposing `Int`-based API and are deferring defining `Index` and `Iterator` until more of the non-escapable collection story is sorted out. The below is some of our research into different potential designs of an `Index` type.

There are 3 potentially-desirable features of `Span`'s `Index` design:

1. `Span` is its own slice type
2. Indices from a slice can be used on the base collection
3. Additional reuse-after-free checking

Each of these introduces practical tradeoffs in the design.

#### `Span` is its own slice type

Collections which own their storage have the convention of separate slice types, such as `Array` and `String`. This has the advantage of clearly delineating storage ownership in the programming model and the disadvantage of introducing a second type through which to interact.

> When types do not own their storage, separate slice types can be [cumbersome](https://github.com/apple/swift/blob/bcd08c0c9a74974b4757b4b8a2d1796659b1d940/stdlib/public/core/StringComparison.swift#L175). The reason `UnsafeBufferPointer` has a separate slice type is because it wants to allow indices to be reused across slices and its `Index` is a relative offset from the start (`Int`) rather than an absolute position (such as a pointer).

Reviewer: ```suggestion
When types do not own their storage, separate slice types can be [cumbersome](https://github.com/swiftlang/swift/blob/swift-5.10.1-RELEASE/stdlib/public/core/StringComparison.swift#L175). The reason `UnsafeBufferPointer` has a separate slice type is because it wants to allow indices to be reused across slices and its `Index` is a relative offset from the start (`Int`) rather than an absolute position (such as a pointer).
```

let span2 = array2.span
// array2 happens to be allocated within the same memory of array1
// but with a different base address whose offset is not an even
// multiple of `MemoryLayout<T>.stride`.

span2[idx] // misaligned load, what happens?
```

If `T` is `BitwiseCopyable`, then the misaligned load is not undefined behavior, but the value that is loaded is garbage. Whether the program is well-behaved going forwards depends on whether it is resilient to getting garbage values.

If `T` is not `BitwiseCopyable`, then the misaligned load may introduce undefined behavior. No matter how well-written the rest of the program is, it has a critical safety and security flaw.

When the reused allocation happens to be stride-aligned, there is no undefined behavior from undefined loads, nor are there "garbage" values in the strictest sense, but it is still reflective of a programming bug. The program may be interacting with an unexpected value.

Bounds checks protect against critical programmer errors. It would be nice, pending engineering tradeoffs, to also protect against some reuse after free errors and invalid index reuse, especially those that may lead to undefined behavior.

> Future improvements to microarchitecture may make reuse after free checks cheaper, however we need something for the forseeable future. Any validation we can do reduces the need to switch to other mitigation strategies or make other tradeoffs.

Reviewer: ```suggestion
Future improvements to microarchitecture may make reuse after free checks cheaper, however we need something for the foreseeable future. Any validation we can do reduces the need to switch to other mitigation strategies or make other tradeoffs.
```

  // note: additional preconditions apply until the end of the scope
  var storage: Span<Element> { get }
}
extension UnsafeRawBufferPointer {
  // note: additional preconditions apply until the end of the scope
  var storage: Span<UInt8> { get }
}
extension UnsafeMutableRawBufferPointer {
  // note: additional preconditions apply until the end of the scope
  var storage: Span<UInt8> { get }
}
```

#### Using `Span` with C functions or other unsafe code:

`Span` has an unsafe hatch for use with unsafe code.

```swift
extension Span where Element: ~Copyable & ~Escapable {
  func withUnsafeBufferPointer<E: Error, Result: ~Copyable & ~Escapable>(
    _ body: (_ buffer: UnsafeBufferPointer<Element>) throws(E) -> Result
  ) throws(E) -> Result
}

extension Span where Element: BitwiseCopyable {
  func withUnsafeBytes<E: Error, Result: ~Copyable & ~Escapable>(
    _ body: (_ buffer: UnsafeRawBufferPointer) throws(E) -> Result
  ) throws(E) -> Result
}
```

### Complete `Span` API:

```swift
public struct Span<Element: ~Copyable & ~Escapable>: Copyable, ~Escapable {
  internal var _start: UnsafePointer<Element>
  internal var _count: Int
}
```

##### Creating a `Span`:

> The initialization of a `Span` instance from an unsafe pointer is an unsafe operation. When it is initialized correctly, subsequent uses of the borrowed instance are safe. Typically these initializers will be used internally to a container's implementation of functions or computed properties that return a borrowed `Span`.

Reviewer 7: I think it might be more correct to say that not only does `Span` have to be initialized correctly, but the memory backing the `Span` has to be managed correctly until the `Span` is no longer in use. For example, guaranteeing the lifetime. For example, one issue we have seen recently involves `mmap`ing some bytes, something unmapping it, then access to the now-gone values.

Author: I improved this paragraph with a more explicit requirement.

```swift
extension Span where Element: ~Copyable & ~Escapable {
  public var count: Int { get }
  public var isEmpty: Bool { get }

  public subscript(_ position: Int) -> Element { _read }
}
```

##### Accessing subranges of elements:

In SE-0437, `UnsafeBufferPointer`'s slicing subscript was replaced by the `extracting(_ bounds:)` functions, due to the copyability assumption baked into the standard library's `Slice` type. `Span` has similar requirements as `UnsafeBufferPointer`, in that it must be possible to obtain an instance representing a subrange of the same memory as another `Span`. We therefore follow in the footsteps of SE-0437, and add a family of `extracting()` functions:

```swift
extension Span where Element: ~Copyable & ~Escapable {
  /// Constructs a new span over the items within the supplied range of
  /// positions within this span.
  ///
  /// The returned span's first item is always at offset 0; unlike buffer

>   /// slices, extracted spans do not generally share their indices with the

Reviewer 7: What does the word `generally` mean here? From a caller's perspective, I assume they should just fully expect that the index is not shared at all, right?

Author: Yeah that's not very good. I adapted this from the new `extracting` functions on `UnsafeBufferPointer`, which use similar words. If you extract a sub-span with the first N items of a span, then the first N indices are effectively shareable, but that's an obvious coincidence, and we don't need to accommodate it by saying "generally".

  ) -> Result
}

extension Span where Element: BitwiseCopyable {
  func withUnsafeBytes<Result>(
    _ body: (_ buffer: UnsafeRawBufferPointer) -> Result
  ) -> Result
}
```

### Complete `Span` API:

```swift
public struct Span<Element: ~Copyable & ~Escapable>: Copyable, ~Escapable {
  internal var _start: UnsafePointer<Element>
  internal var _count: Int
}
```

##### Creating a `Span`:

The initialization of a `Span` instance from an unsafe pointer is an unsafe operation. When it is initialized correctly, subsequent uses of the borrowed instance are safe. Typically these initializers will be used internally to a container's implementation of functions or computed properties that return a borrowed `Span`.

```swift
extension Span {

  /// Unsafely create a `Span` over initialized memory.
  ///
  /// The memory in `buffer` must be owned by the instance `owner`,
  /// meaning that as long as `owner` is alive the memory will remain valid.
  ///
  /// - Parameters:
  ///   - buffer: an `UnsafeBufferPointer` to initialized elements.
  ///   - owner: a binding whose lifetime must exceed that of
  ///            the newly created `Span`.

>   public init<Owner>(

Reviewer 2: ```suggestion
  public init<Owner: ~Copyable>(
```

Should it also be marked `~Escapable`?

Author: All the owners have been marked `~Copyable & ~Escapable`

    unsafeElements buffer: UnsafeBufferPointer<Element>,
    owner: borrowing Owner
  ) -> dependsOn(owner) Self

  /// Unsafely create a `Span` over initialized memory.
  ///
  /// The memory representing `count` instances starting at
  /// `pointer` must be owned by the instance `owner`,
  /// meaning that as long as `owner` is alive the memory will remain valid.
  ///
  /// - Parameters:
  ///   - pointer: a pointer to the first initialized element.
  ///   - count: the number of initialized elements in the view.
  ///   - owner: a binding whose lifetime must exceed that of
  ///            the newly created `Span`.

>   public init<Owner>(

Reviewer 2: ```suggestion
  public init<Owner: ~Copyable>(
```

    unsafeStart pointer: UnsafePointer<Element>,
    count: Int,
    owner: borrowing Owner
  ) -> dependsOn(owner) Self
}

extension Span where Element: BitwiseCopyable {

  /// Unsafely create a `Span` over initialized memory.
  ///
  /// The memory in `unsafeBytes` must be owned by the instance `owner`
  /// meaning that as long as `owner` is alive the memory will remain valid.
  ///
  /// `unsafeBytes` must be correctly aligned for accessing
  /// an element of type `Element`, and must contain a number of bytes
  /// that is an exact multiple of `Element`'s stride.
  ///
  /// - Parameters:
  ///   - unsafeBytes: a buffer to initialized elements.
  ///   - type: the type to use when interpreting the bytes in memory.
  ///   - owner: a binding whose lifetime must exceed that of
  ///            the newly created `Span`.

>   public init<Owner>(

Reviewer 2: ```suggestion
  public init<Owner: ~Copyable>(
```

    unsafeBytes buffer: UnsafeRawBufferPointer,
    owner: borrowing Owner
  ) -> dependsOn(owner) Self

  /// Unsafely create a `Span` over a span of initialized memory.
  ///
  /// The memory representing `count` instances starting at
  /// `unsafeRawPointer` must be owned by the instance `owner`,
  /// meaning that as long as `owner` is alive the memory will remain valid.
  ///
  /// `unsafeRawPointer` must be correctly aligned for accessing
  /// an element of type `Element`.
  ///
  /// - Parameters:
  ///   - unsafeRawPointer: a pointer to the first initialized element.
  ///   - type: the type to use when interpreting the bytes in memory.
  ///   - count: the number of initialized elements in the view.
  ///   - owner: a binding whose lifetime must exceed that of
  ///            the newly created `Span`.

>   public init<Owner>(

Reviewer 2: ```suggestion
  public init<Owner: ~Copyable>(
```

  ///
  /// - Parameters:
  ///   - span: An existing `Span<T>`, which will define both this
  ///           `RawSpan`'s lifetime and the memory it represents.
  @inlinable @inline(__always)
  public init<T: BitwiseCopyable>(
    _ span: borrowing Span<T>
  ) -> dependsOn(span) Self
}
```

##### Accessing the memory of a `RawSpan`:

The basic operations to access the contents of the memory underlying a `RawSpan` are `load(as:)` and `loadUnaligned(as:)`.

```swift
extension RawSpan {
  /// Returns a new instance of the given type, constructed from the raw memory
  /// at the specified offset.
  ///
  /// The memory at this pointer plus `offset` must be properly aligned for
  /// accessing `T` and initialized to `T` or another type that is layout
  /// compatible with `T`.
  ///
  /// - Parameters:
  ///   - offset: The offset from this pointer, in bytes. `offset` must be
  ///     nonnegative. The default is zero.
  ///   - type: The type of the instance to create.
  /// - Returns: A new instance of type `T`, read from the raw bytes at
  ///     `offset`. The returned instance is memory-managed and unassociated
  ///     with the value in the memory referenced by this pointer.

>   public func load<T>(

Reviewer 2: ```suggestion
  public func load<T: ~Copyable>(
```

    fromByteOffset offset: Int = 0, as: T.Type
  ) -> T

  /// Returns a new instance of the given type, constructed from the raw memory
  /// at the specified offset.
  ///
  /// The memory at this pointer plus `offset` must be properly aligned for
  /// accessing `T` and initialized to `T` or another type that is layout
  /// compatible with `T`.
  ///
  /// This function does not validate the bounds of the memory access;
  /// this is an unsafe operation.
  ///
  /// - Parameters:
  ///   - offset: The offset from this pointer, in bytes. `offset` must be
  ///     nonnegative. The default is zero.
  ///   - type: The type of the instance to create.
  /// - Returns: A new instance of type `T`, read from the raw bytes at
  ///     `offset`. The returned instance is memory-managed and unassociated
  ///     with the value in the memory referenced by this pointer.

>   public func load<T>(

Reviewer 2: ```suggestion
  public func load<T: ~Copyable>(
```

The `UnsafeBufferPointer` passed to a `withUnsafeXXX` closure-style API, while performant, is unsafe in multiple ways:

1. The pointer itself is unsafe and unmanaged
2. `subscript` is only bounds-checked in debug builds of client code
3. It might escape the duration of the closure

Even if the body of the `withUnsafeXXX` call does not escape the pointer, other functions called inside the closure have to be written in terms of unsafe pointers. This requires programmer vigilance across a project and potentially spreads the use of unsafe types, even when it could have been written in terms of safe constructs.

We want to take advantage of the features of non-escapable types to replace some closure-taking API with simple properties, resulting in more composable code:

> ```swift
> let array = Array("Hello\0".utf8)
> array.withUnsafeBufferPointer {
>   // use `$0` here for direct memory access
> }
> let span: Span<UInt8> = array.storage
> // use `span` in the same scope as `array` for direct memory access
> ```

Reviewer 8: ```suggestion
```swift
let array = Array("Hello\0".utf8)

// Old
array.withUnsafeBufferPointer {
  // use `$0` here for direct memory access
}

// New
let span: Span<UInt8> = array.storage
// use `span` in the same scope as `array` for direct memory access
```
```

`Span` is the currency type for local processing over values in contiguous memory. It is the replacement for any API currently using `Array`, `UnsafeBufferPointer`, `Foundation.Data`, etc., that does not need to escape the value.

### `ContiguousStorage`

A type can indicate that it can provide a `Span` by conforming to the `ContiguousStorage` protocol. `ContiguousStorage` forms a bridge between multi-type or generically-typed interfaces and a performant concrete implementation.

For example, for the hypothetical base64 decoding library mentioned above, a possible API could be:

```swift
extension HypotheticalBase64Decoder {
  public func decode(bytes: some ContiguousStorage<UInt8>) -> [UInt8]
}
```

Even better, an interface can be defined in terms of the concrete type `Span<UInt8>`:

```swift
extension Hypothetical Base64Decoder {
  public func decode(bytes: Span<UInt8>) -> [UInt8]
}
```

> Advanced libraries might add use an inlinable generic-dispatch interface in addition to a concrete interface defined in terms of `Span`

Reviewer 8: We say "even better" for `Span` vs `ContiguouslyStored`, but don't directly prescribe this. We can insert some of the prose that recommends using `Span` in most places and leaving `ContiguouslyStored` as adapter code or in select top-level API. Since `ContiguouslyStored` exposes a `.storage` property, it's not a huge obstacle.

### `RawSpan`

`RawSpan` allows sharing the contiguous internal representation for values which may be heterogenously-typed, such as in decoders. Since it is a fully concrete type, it can achieve better performance in debug builds of client code as well as a more straight-forwards understanding of performance for library code.

`Span<some BitwiseCopyable>` can always be converted to `RawSpan`, using a conditionally-available property or a constructor.

## Detailed design

`Span<Element>` is a simple representation of a span of initialized memory.

```swift

> public struct Span<Element: ~Copyable & ~Escapable>: Copyable, ~Escapable {

Reviewer 8: Are we pitching `~Escapable` `Element`?

Author: I don't want to foreclose it until we must. We might not have it in the reviewed proposal, but for now it can stay.

Reviewer 3: Reminder: the underlying pointer needs to be `UnsafeRawPointer`, like it was in the BufferView prototypes. A `Span<T: BitwiseCopyable>` can be reinterpreted as `Span<U: BitwiseCopyable>`. `RawSpan` is fine but it doesn't actually solve any problems.

  internal var _start: UnsafePointer<Element>
  internal var _count: Int
}
```

It provides a buffer-like interface to the elements stored in that span of memory:

```swift
extension Span {
  public var count: Int { get }
  public var isEmpty: Bool { get }

>   public var indices: Range<Int> { get }

Reviewer 8: I think this is well-motivated by the `for` loop, but I will note that this might preclude us from formally defining an `Index` type in the future depending on how non-escapable collections go. We could also defer this.

Author: We can give it a different name, such as `offsets`, but it feels like using `indices` for _this_ type isn't particularly risky.

Author: I'll remove `indices` from the proposal. In the prototype it's now `_indices` and `_byteOffsets` for `Span` and `RawSpan` respectively.

>   subscript(_ position: Int) -> Element { get }

Reviewer 2: The accessor `get` here does not work for `Span`; we need this document to either admit to the existence of read accessors, or to leave it unspecified for now, explicitly deferring this detail to a future proposal.

Author: I declared `_read` accessors where appropriate, and added a section about future coroutine accessors to future directions.

}
```

Note that `Span` does _not_ conform to `Collection`. This is because `Collection`, as originally conceived and enshrined in existing source code, assumes pervasive copyability and escapability for itself as well as its elements. In particular a subsequence of a `Collection` is semantically a separate value from the instance it was derived from. In the case of `Span`, a sub-span representing a subrange of its elements _must_ have the same lifetime as the view from which it originates. Another proposal will consider collection-like protocols to accommodate different combinations of `~Copyable` and `~Escapable` for the collection and its elements.

`Span`s representing subsets of consecutive elements can be extracted out of a larger `Span` with an API similar to the recently added `extracting()` functions of `UnsafeBufferPointer`:

```swift
extension Span {
  public func extracting(_ bounds: Range<Int>) -> Self
}
```

> 

Reviewer 8: To me, subjectively, subscript reads better.  SE-0437's justification was because we don't have `Slice<~Copyable>`, but that doesn't apply if we don't have a slice type. 

Perhaps our rationale is that we want to reserve the right to a different slice type in the future.

Author: I agree that subscript reads better, but the stdlib's pattern is that what's returned by the range subscript getter shares its indices with the original binding. It feels wrong to make something that behaves so differently while looking the same. Discussions welcome.

Reviewer 2: The slicing subscript notation would indeed read better. However, Swift has very strongly established the expectation that the result of `foo[i ..< j]` will share indices with the original container, which is not the case here. We should not use the same notation to mean a very different thing; especially not in the context of a shape that is so close to our preexisting primitive container types.

Additionally, the slicing subscript notation we have in classic Swift is able to accommodate all three flavors of access (borrowing, mutating and consuming) with a single, universal result type. This isn't going to be possible within the ownership model -- we'll need to invent distinct notations for extracting borrowing vs mutating parts. I don't think `Span` is the right place to establish a general pattern for this matter.

The first element of a given span is _always_ at position zero, and its last it as position `count-1`.

As a side-effect of not conforming to `Collection` or `Sequence`, `Span` is not directly supported by `for` loops at this time. It is, however, easy to use in a `for` loop via indexing:

```swift

> for i in mySpan.indices {

Reviewer 2: I do not think we want to establish an expectation that noncopyable containers would come with a property called `indices`. Direct for-in loops over spans are expected to come very soon, in the proposal that introduces borrowing sequences. We do not need to bend over backwards to make this nice right now; I think we especially do not need to add any API to cater for it.

```suggestion
for i in 0 ..<  mySpan.count {
```

Author: Did this independently =)

  calculation(mySpan[i])
}
```

### `ContiguousStorage`

A type can declare that it can provide access to contiguous storage by conforming to the `ContiguousStorage` protocol:

```swift
public protocol ContiguousStorage<Element>: ~Copyable, ~Escapable {
  associatedtype Element: ~Copyable & ~Escapable

>   var storage: Span<Element> { get }

Reviewer 2: Same issue with `get` here as we have with `Span`'s subscript. This establishes a protocol requirement; as I understand it, we actually need it to require a read accessor, which we cannot spell yet.

Author: The requirement for a `_read` accessor is only one problem for protocols. The other is that we actually cannot suppress requirements on `associatedtype` declarations yet. (This capability was deferred late during the Noncopyable Generics review.)

Author: For the reasons above, I've moved `ContiguousStorage` to "Future Directions".

}
```

The key safety feature is that a `Span` cannot escape to a scope where the value it borrowed no longer exists.

A function that wishes to read from contiguous storage can declare a parameter type of `some ContiguousStorage`. The implementation will internally consist of a brief generic section, followed by business logic implemented in terms of a concrete `Span`. Frameworks that support library evolution (resilient frameworks) have an additional concern. Resilient frameworks have an ABI boundary that may differ from the API proper. Resilient frameworks may wish to adopt a pattern such as the following:

```swift
extension MyResilientType {
  // public API
  @inlinable public func essentialFunction(_ a: some ContiguousStorage<some Any>) -> Int {
    self.essentialFunction(a.storage)
  }

  // ABI boundary
  public func essentialFunction(_ a: Span<some Any>) -> Int { ... }
}
```

> Here, the public function obtains the `Span` from the type that vends it in inlinable code, then calls a concrete, opaque function defined in terms of `Span`. Inlining the generic shim in the client is often a critical optimization. The need for such a pattern and related improvements are discussed in the future directions below (see [Syntactic Sugar for Automatic Conversions](#Conversions).)

Reviewer 2: For the record, I remain very much unconvinced that it'll be wise to encourage folks to create APIs that can only operate on contiguous storage, as opposed to a series of contiguous storage chunks.

  var storage: Span<Element> { get }
}

extension Foundation.Data: ContiguousStorage<UInt8> {
  var storage: Span<UInt8> { get }
}

extension String.UTF8View: ContiguousStorage<Unicode.UTF8.CodeUnit> {
  // note: this could borrow a temporary copy of the `String`'s storage
  var storage: Span<Unicode.UTF8.CodeUnit> { get }
}
extension Substring.UTF8View: ContiguousStorage<Unicode.UTF8.CodeUnit> {
  // note: this could borrow a temporary copy of the `Substring`'s storage
  var storage: Span<Unicode.UTF8.CodeUnit> { get }
}
extension Character.UTF8View: ContiguousStorage<Unicode.UTF8.CodeUnit> {
  // note: this could borrow a temporary copy of the `Character`'s storage
  var storage: Span<Unicode.UTF8.CodeUnit> { get }
}

extension SIMD: ContiguousStorage<Self.Scalar> {
  var storage: Span<Self.Scalar> { get }
}
extension KeyValuePairs: ContiguousStorage<(Self.Key, Self.Value)> {
  var storage: Span<(Self.Key, Self.Value)> { get }
}
extension CollectionOfOne: ContiguousStorage<Element> {
  var storage: Span<Element> { get }
}

extension Slice: ContiguousStorage where Base: ContiguousStorage {
  var storage: Span<Base.Element> { get }
}
```

> In addition to the the safe types above gaining the `storage` property, the `UnsafeBufferPointer` family of types will also gain access to a `storage` property. This enables interoperability if `Span`-taking API. While a `Span` binding created from an `UnsafeBufferPointer` exists, the memory that underlies it must not be deinitialized or deallocated.

Reviewer 2: This seems like a bad idea to me, because it encourages accidents. 

The right way to convert an UBP to a Span is to use one of the initializers that take an owner.

Author: I'll remove this. I've become less convinced myself

Author: Removed in favor of an example directly using a `Span` initializer.

> In addition to the the safe types above gaining the `storage` property, the `UnsafeBufferPointer` family of types will also gain access to a `storage` property. This enables interoperability if `Span`-taking API. While a `Span` binding created from an `UnsafeBufferPointer` exists, the memory that underlies it must not be deinitialized or deallocated.
> 

Reviewer 8: ```suggestion
In addition to the the safe types above gaining the `storage` property, the `UnsafeBufferPointer` family of types will also gain access to a `storage` property. This enables interoperability of `Span`-taking API. While a `Span` binding created from an `UnsafeBufferPointer` exists, the memory that underlies it must not be deinitialized or deallocated.

```

```swift
extension UnsafeBufferPointer: ContiguousStorage<Self.Element> {
  // note: additional preconditions apply until the end of the scope
  var storage: Span<Element> { get }
}
extension UnsafeMutableBufferPointer: ContiguousStorage<Self.Element> {
  // note: additional preconditions apply until the end of the scope
  var storage: Span<Element> { get }
}
extension UnsafeRawBufferPointer: ContiguousStorage<UInt8> {
  // note: additional preconditions apply until the end of the scope
  var storage: Span<UInt8> { get }
}
extension UnsafeMutableRawBufferPointer: ContiguousStorage<UInt8> {
  // note: additional preconditions apply until the end of the scope
  var storage: Span<UInt8> { get }
}
```

> 

Reviewer 8: I'm not 100% sure we want to do this generically. Elsewhere, we have an argument label `unsafe` when constructing a `Span` from unsafe pointers. 

This is unsafe-to-safe adapter code and we have explicit inits for those.

Author: That is fair.

Author: Removed in favor of an example directly using a Span initializer.

#### Using `Span` with C functions or other unsafe code:

`Span` has an unsafe hatch for use with unsafe code.

```swift
extension Span {
  func withUnsafeBufferPointer<Result>(
    _ body: (_ buffer: UnsafeBufferPointer<Element>) -> Result
  ) -> Result
}

extension Span where Element: BitwiseCopyable {
  func withUnsafeBytes<Result>(
    _ body: (_ buffer: UnsafeRawBufferPointer) -> Result
  ) -> Result
}
```

### Complete `Span` API:

```swift
public struct Span<Element: ~Copyable & ~Escapable>: Copyable, ~Escapable {
  internal var _start: Span<Element>.Index
  internal var _count: Int
}
```

##### Creating a `Span`:

> The initialization of a `Span` instance is an unsafe operation. When it is initialized correctly, subsequent uses of the borrowed instance are safe. Typically these initializers will be used internally to a container's implementation of functions or computed properties that return a borrowed `Span`.

Reviewer 8: ```suggestion
The initialization of a `Span` instance from an unsafe pointer is an unsafe operation. When it is initialized correctly, subsequent uses of the borrowed instance are safe. Typically these initializers will be used internally to a container's implementation of functions or computed properties that return a borrowed `Span`.
```

```swift
extension Span {

  /// Unsafely create a `Span` over initialized memory.
  ///
  /// The memory in `buffer` must be owned by the instance `owner`,
  /// meaning that as long as `owner` is alive the memory will remain valid.

>   ///

Reviewer 8: Returns `nil` if `buffer.baseAddress` is `nil`.

Author: @lorentey would prefer this to not be failable, fwiw.

Author: We will make this non-failable.

Reviewer 8: Would it trap on nil or make an empty span on nil (and would base address he 0?)?

Reviewer 2: It would need to conjur a pointer value out of thin air. We can define an empty singleton for this purpose (say, a dummy UInt8), and put it in a read-only section.

Reviewer 2: (Trapping and returning nil both have the fatal flaw that they force the client to care about whether their empty buffer happens to have a nil base. That would be terrible -- the client isn't necessarily in a position to dictate that, and forcing them to try to somehow manually handle the nil case would only lead to entirely avoidable bugs.)

Author: There are no failable initializers anymore in the proposal. The implementation still traps when `baseAddress` is `nil`; I think we can defer conjuring a pointer until adapting the work in the standard library.

  ///   - owner: a binding whose lifetime must exceed that of
  ///            the newly created `Span`.
  public init<Owner>(
    unsafeBytes buffer: UnsafeRawBufferPointer,
    owner: borrowing Owner
  ) -> dependsOn(owner) Self

  /// Unsafely create a `Span` over a span of initialized memory.
  ///
  /// The memory representing `count` instances starting at
  /// `unsafeRawPointer` must be owned by the instance `owner`,
  /// meaning that as long as `owner` is alive the memory will remain valid.
  ///
  /// `unsafeRawPointer` must be correctly aligned for accessing
  /// an element of type `Element`.
  ///
  /// - Parameters:
  ///   - unsafeRawPointer: a pointer to the first initialized element.
  ///   - type: the type to use when interpreting the bytes in memory.
  ///   - count: the number of initialized elements in the view.
  ///   - owner: a binding whose lifetime must exceed that of
  ///            the newly created `Span`.
  public init<Owner>(
    unsafeStart pointer: UnsafeRawPointer,
    byteCount: Int,
    owner: borrowing Owner
  ) -> dependsOn(owner) Self
}
```

#####  `Collection`-like API:

> The following typealiases, properties, functions and subscripts have direct counterparts in the `Collection` protocol hierarchy. Their semantics shall be as described where they counterpart is declared (in `Collection` or `RandomAccessCollection`). The only difference with their counterpart should be a lifetime dependency annotation where applicable, allowing them to return borrowed nonescapable values or borrowed noncopyable values.

Reviewer 8: ```suggestion
The following properties, functions and subscripts have direct counterparts in the `Collection` protocol hierarchy. Their semantics shall be as described where they counterpart is declared (in `Collection` or `RandomAccessCollection`). The only difference with their counterpart should be a lifetime dependency annotation where applicable, allowing them to return borrowed nonescapable values or borrowed noncopyable values.
```

```swift

> extension Span {

Reviewer 2: ```suggestion
extension Span where Element: ~Copyable {
```

Author: oh hey, that's every single `Span` extension in the document.

Author: fixed the `Span` extensions to all either support non-copyable element types or require `BitwiseCopyable` elements.

  public var count: Int { get }
  public var isEmpty: Bool { get }

>   public var indices: Range<Int> { get }

Reviewer 2: ```suggestion
```

Author: already removed

>   public subscript(_ position: Int) -> dependsOn(self) Element { get }

Reviewer 2: What's the purpose of `dependOn` modifier here?

Reviewer 2: There is also the issue of `get`.

Author: I want to clean up the `dependsOn` modifiers that are just the defaults. (That might be all of them.) At an earlier time there was no inference for the lifetime dependencies, and a bunch of the explicitness remains.

```swift
extension Span {
  func extracting(_ bounds: Range<Int>) -> dependsOn(self) Self
  func extracting(_ bounds: some RangeExpression<Int>) -> dependsOn(self) Self
  func extracting(_: UnboundedRange) -> dependsOn(self) Self
}
```

##### Unchecked access to elements and subranges of elements:

The `subscript` and the `extracting()` functions mentioned above all have always-on bounds checking of their parameters, in order to prevent out-of-bounds accesses. We also want to provide unchecked variants as an alternative for cases where bounds-checking is proving costly.

```swift
extension Span {
  // Unchecked subscripting and extraction

  /// Accesses the element at the specified `position`.
  ///
  /// This subscript does not validate `position`; this is an unsafe operation.
  ///
  /// - Parameter position: The position of the element to access. `position`
  ///     must be a valid index that is not equal to the `endIndex` property.
  public subscript(
    unchecked position: Index

>   ) -> dependsOn(self) Element { get }

Reviewer 2: Same question about `dependsOn(self)` and `get`.

Author: I audited the `dependsOn` annotations, and it turns out none of them were needed (so they're removed.)
@tbkka believes we should actually mention `_read` accessors, so I will.

  /// Accesses a contiguous subrange of the elements represented by this `Span`
  ///
  /// This subscript does not validate `bounds`; this is an unsafe operation.
  ///
  /// - Parameter bounds: A range of the collection's indices. The bounds of
  ///     the range must be valid indices of the collection.

>   public extracting(

Reviewer 2: ```suggestion
  public func extracting(
```

    uncheckedBounds bounds: Range<Index>
  ) -> dependsOn(self) Self

  /// Accesses the contiguous subrange of the elements represented by
  /// this `Span`, specified by a range expression.
  ///
  /// This subscript does not validate `bounds`; this is an unsafe operation.
  ///
  /// - Parameter bounds: A range of the collection's indices. The bounds of
  ///     the range must be valid indices of the collection.

>   public subscript(

Reviewer 8: This is `subscript`, not `extracting`.

Author: Fixed

>   public subscript(

Author: ```suggestion
  public func extracting(
```

    uncheckedBounds bounds: some RangeExpression<Index>
  ) -> dependsOn(self) Span<Element>
}
```

##### Index validation utilities:

Every time `Span` uses a position parameter, it checks for its validity, unless the parameter is marked with the word "unchecked". The validation is performed with these functions:

```swift
extension Span {
  /// Traps if `position` is not a valid offset into this `Span`
  ///
  /// - Parameters:
  ///   - position: an position to validate

>   public boundsCheckPrecondition(_ position: Index)

Author: ```suggestion
  public boundsCheckPrecondition(_ position: Int)
```

>   public boundsCheckPrecondition(_ position: Index)

Reviewer 2: These aren't good names. Do we really want these to trap rather than returning a Bool?

Author: We want the trapping one internally, and I don't see a reason not to expose them, since a wrapper might be interested in using them. We should have the `Bool`-returning validating ones too.

Author: Added a `Bool`-returning bounds-checking function in addition to the (renamed) trapping one.

  /// Traps if `bounds` is not a valid range of offsets into this `Span`
  ///
  /// - Parameters:
  ///   - position: a range of positions to validate

>   public boundsCheckPrecondition(_ bounds: Range<Index>)

Author: ```suggestion
  public boundsCheckPrecondition(_ bounds: Range<Int>)
```

Author: Also missing a `func` keyword. Added.

}

> ```

Reviewer 8: These talk about `Index`, but IIRC we don't define `Index`.

Author: Found and replaced the remaining instances of the word Index

Reviewer 2: We should define `Index`! As in, we should literally have `typealias Index = Int` in `Span`.

(SE-0437 has sunk the concept of `Index` down to the `~Copyable` world for the UnsafeBufferPointer types.)

##### Interoperability with unsafe code:

We provide two functions for interoperability with C or other legacy pointer-taking functions.

```swift
extension Span {
  /// Calls a closure with a pointer to the viewed contiguous storage.
  ///
  /// The buffer pointer passed as an argument to `body` is valid only
  /// during the execution of `withUnsafeBufferPointer(_:)`.
  /// Do not store or return the pointer for later use.
  ///
  /// - Parameter body: A closure with an `UnsafeBufferPointer` parameter
  ///   that points to the viewed contiguous storage. If `body` has
  ///   a return value, that value is also used as the return value
  ///   for the `withUnsafeBufferPointer(_:)` method. The closure's
  ///   parameter is valid only for the duration of its execution.
  /// - Returns: The return value of the `body` closure parameter.

>   func withUnsafeBufferPointer<Result>(
>     _ body: (_ buffer: UnsafeBufferPointer<Element>) -> Result
>   ) -> Result

Reviewer 2: ```suggestion
  func withUnsafeBufferPointer<E: Error, Result: ~Copyable>(
    _ body: (_ buffer: UnsafeBufferPointer<Element>) throws(E) -> Result
  ) throws(E) -> Result
```

}

extension Span where Element: BitwiseCopyable {
  /// Calls the given closure with a pointer to the underlying bytes of
  /// the viewed contiguous storage.
  ///
  /// The buffer pointer passed as an argument to `body` is valid only
  /// during the execution of `withUnsafeBytes(_:)`.
  /// Do not store or return the pointer for later use.
  ///
  /// - Parameter body: A closure with an `UnsafeRawBufferPointer`
  ///   parameter that points to the viewed contiguous storage.
  ///   If `body` has a return value, that value is also
  ///   used as the return value for the `withUnsafeBytes(_:)` method.
  ///   The closure's parameter is valid only for the duration of
  ///   its execution.
  /// - Returns: The return value of the `body` closure parameter.

>   func withUnsafeBytes<Result>(
>     _ body: (_ buffer: UnsafeRawBufferPointer) -> Result
>   ) -> Result

Reviewer 2: ```suggestion
  func withUnsafeBytes<E: Error, Result: ~Copyable>(
    _ body: (_ buffer: UnsafeRawBufferPointer) throws(E) -> Result
  ) throws(E) -> Result
```

}
```

### RawSpan

In addition to `Span<T>`, we propose the addition of `RawSpan`, to represent heterogenously-typed values in contiguous memory. `RawSpan` is similar to `Span<T>`, but represents _untyped_ initialized bytes. `RawSpan` is a specialized type that intends to support parsing and decoding applications, as well as applications where heavily-used code paths require concrete types as much as possible. Its API supports extracting sub-spans, along with the operations `load(as:)` and `loadUnaligned(as:)`.

#### Complete `RawSpan` API:

```swift
public struct RawSpan: Copyable, ~Escapable {

>   internal var _start: RawSpan.Index

Reviewer 8: We don't have `Index`.

Author: Found and replaced the last instances of the word `Index`

>   internal var _start: RawSpan.Index

Author: ```suggestion
  internal var _start: UnsafeRawPointer
```

  internal var _count: Int
}
```

##### Initializing a `RawSpan`:

```swift
extension RawSpan {
  /// Unsafely create a `RawSpan` over initialized memory.
  ///
  /// The memory in `buffer` must be owned by the instance `owner`,
  /// meaning that as long as `owner` is alive the memory will remain valid.
  ///
  /// - Parameters:
  ///   - buffer: an `UnsafeRawBufferPointer` to initialized memory.
  ///   - owner: a binding whose lifetime must exceed that of
  ///            the newly created `RawSpan`.
  public init?<Owner: ~Copyable & ~Escapable>(

>     unsafeBytes buffer: UnsafeBufferPointer<Element>,

Reviewer 8: Does `Element` need to be bitwise copyable? Everywhere else we do this kind of conversion we have that constraint.

Author: This one was supposed to be `UnsafeRawBufferPointer`!
```suggestion
    unsafeBytes buffer: UnsafeRawBufferPointer,
```

    owner: borrowing Owner
  ) -> dependsOn(owner) Self?

  /// Unsafely create a `RawSpan` over initialized memory.
  ///
  /// The memory over `count` bytes starting at
  /// `pointer` must be owned by the instance `owner`,
  /// meaning that as long as `owner` is alive the memory will remain valid.
  ///
  /// - Parameters:
  ///   - pointer: a pointer to the first initialized element.
  ///   - count: the number of initialized elements in the view.
  ///   - owner: a binding whose lifetime must exceed that of
  ///            the newly created `RawSpan`.

>   public init<Owner: ~Copyable & ~Escapable>(
>     unsafeRawPointer pointer: UnsafeRawPointer,
>     count: Int,
>     owner: borrowing Owner
>   ) -> dependsOn(owner) Self

Reviewer 2: ```suggestion
  public init<Owner: ~Copyable & ~Escapable>(
    unsafeStart pointer: UnsafeRawPointer,
    byteCount: Int,
    owner: borrowing Owner
  ) -> dependsOn(owner) Self
```

Author: This has been changed

  /// Create a `RawSpan` over the memory represented by a `Span<T>`
  ///
  /// - Parameters:
  ///   - span: An existing `Span<T>`, which will define both this
  ///           `RawSpan`'s lifetime and the memory it represents.

>   @inlinable @inline(__always)

Reviewer 2: Nit: `@inline(__always)` is an implementation detail; it should not appear as part of the proposed API.

Author: Yeah that's a copy-pasting error
```suggestion
```

  /// this is an unsafe operation.
  ///
  /// - Parameters:
  ///   - offset: The offset from this pointer, in bytes. `offset` must be
  ///     nonnegative. The default is zero.
  ///   - type: The type of the instance to create.
  /// - Returns: A new instance of type `T`, read from the raw bytes at
  ///     `offset`. The returned instance isn't associated
  ///     with the value in the range of memory referenced by this pointer.
  public func loadUnaligned<T: BitwiseCopyable>(
    fromUncheckedByteOffset offset: Int, as: T.Type
  ) -> T
}
```

A `RawSpan` can be viewed as a `Span<T>`, provided the memory is laid out homogenously as instances of `T`.

```swift
  /// View the memory span represented by this view as a different type
  ///
  /// The memory must be laid out identically to the in-memory representation of `T`.
  ///
  /// - Parameters:
  ///   - type: The type you wish to view the memory as
  /// - Returns: A new `Span` over elements of type `T`
  public func view<T: BitwiseCopyable>(as: T.Type) -> dependsOn(self) Span<T>
}
```

##### Index-related Operations:

```swift
extension RawSpan {
  /// The number of bytes in the span.

>   public var count: Int { get }

Reviewer 2: ```suggestion
  public var byteCount: Int { get }
```

  /// A Boolean value indicating whether the span is empty.
  public var isEmpty: Bool { get }

>   /// The indices that are valid for subscripting the span, in ascending
>   /// order.
>   public var indices: Range<Int> { get }
> 

Reviewer 2: ```suggestion
```

Author: Removed

  ///
  /// - Parameters:
  ///   - offsets: a range of offsets to validate
  public func boundsCheckPrecondition(_ offsets: Range<Int>)
}
```

##### Index validation utiliities:

Every time `RawSpan` uses an index or an integer offset, it checks for their validity, unless the parameter is marked with the word "unchecked". The validation is performed with these functions:

```swift
extension RawSpan {
  /// Traps if `position` is not a valid index for this `RawSpan`
  ///
  /// - Parameters:
  ///   - position: an Index to validate
  public boundsCheckPrecondition(_ position: Index)

  /// Traps if `bounds` is not a valid range of indices for this `RawSpan`
  ///
  /// - Parameters:
  ///   - bounds: a range of indices to validate
  public boundsCheckPrecondition(_ bounds: Range<Index>)

  /// Traps if `offset` is not a valid offset into this `RawSpan`
  ///
  /// - Parameters:
  ///   - offset: an offset to validate
  public boundsCheckPrecondition(offset: Int)

  /// Traps if `offsets` is not a valid range of offsets into this `RawSpan`
  ///
  /// - Parameters:
  ///   - offsets: a range of offsets to validate
  public boundsCheckPrecondition(offsets: Range<Int>)
}

> ```

Reviewer 8: This section is redundant with the functions in the prior one

Author: I'll remove this.

##### Accessing subranges of elements:

Similarly to `Span`, `RawSpan` does not support slicing in the style of `Collection`. It supports a similar set of `extracting()` functions as `Span`:

```swift
extension RawSpan {
  /// Constructs a new span over the bytes within the supplied range of
  /// positions within this span.
  ///
  /// The returned span's first byte is always at offset 0; unlike buffer
  /// slices, extracted spans do not generally share their indices with the
  /// span from which they are extracted.
  ///
  /// - Parameter bounds: A valid range of positions. Every position in
  ///     this range must be within the bounds of this `RawSpan`.
  ///
  /// - Returns: A `Span` over the bytes within `bounds`
  public func extracting(_ bounds: Range<Int>) -> Self

  /// Constructs a new span over the bytes within the supplied range of
  /// positions within this span.
  ///
  /// The returned span's first byte is always at offset 0; unlike buffer
  /// slices, extracted spans do not generally share their indices with the
  /// span from which they are extracted.
  ///

>   /// This function does not validate `bounds`; this is an unsafe operation.

Reviewer 2: General note: We need to avoid guaranteeing that no validation will take place, as we will still need these to fully validate their inputs on certain platforms and/or configurations. Such promises should be weakened accordingly.

```suggestion
  /// This function may not always validate `bounds`; this is an unsafe operation.
```

  /// the span, the result is an empty span.
  ///
  /// - Parameter k: The number of bytes to drop off the end of
  ///   the span. `k` must be greater than or equal to zero.
  /// - Returns: A span leaving off the specified number of bytes at the end.
  public func extracting(droppingLast k: Int) -> Self

  /// Returns a span containing the trailing bytes of the span,
  /// up to the given maximum length.
  ///
  /// If the maximum length exceeds the length of this span,
  /// the result contains all the bytes.
  ///
  /// - Parameter maxLength: The maximum number of bytes to return.
  ///   `maxLength` must be greater than or equal to zero.
  /// - Returns: A span with at most `maxLength` bytes.
  public func extracting(last maxLength: Int) -> Self

  /// Returns a span over all but the given number of initial bytes.
  ///
  /// If the number of elements to drop exceeds the number of bytes in
  /// the span, the result is an empty span.
  ///
  /// - Parameter k: The number of bytes to drop from the beginning of
  ///   the span. `k` must be greater than or equal to zero.
  /// - Returns: A span starting after the specified number of bytes.

>   public func extracting(droppingFirst k: Int = 1) -> Self

Reviewer 8: Should we have these convenience overloads on Span too?

Author: They are intended to be on `Span` too; if I forgot to add them to the proposal I will fix that.

Reviewer 2: I'd argue to drop them entirely, and to try to live with range expressions only.

(If it turns out we do need some of them, we should add them on all related types, including the buffer pointer family.)

Author: Let's see if they get feedback in the updated pitch (I'm curious), and we can skip them when it comes up for review.

}
```

### Byte parsing helpers

> 

Reviewer 8: NOTE: I need to rewrite some of this

Author: Thanks

> The below (severable) API make `RawSpan` well-suited for use in binary parsers and decoders.

Reviewer 2: These additions seem a bit premature to consider adding to the core stdlib. Can we instead prove their worth in a package?

I'm particularly scared by using the name `Cursor` here to mean a borrowing iterator.

Author: They're likely to get used in the parser package, so _that_ might be the right spot.

##### Use a non-escapable index type
Eventually we want a similar usage pattern for a `MutableSpan` as we are proposing for `Span`. If the index of a `MutableSpan` were to borrow the view, then it becomes impossible to implement a mutating subscript without also requiring an index to be consumed. This seems untenable.

##### Naming

The ideas in this proposal previously used the name `BufferView`. While the use of the word "buffer" would be consistent with the `UnsafeBufferPointer` type, it is nevertheless not a great name, since "buffer" is usually used in reference to transient storage. On the other hand we already have a nomenclature using the term "Storage" in the `withContiguousStorageIfAvailable()` function, and we tried to allude to that in a previous pitch where we called this type `StorageView`. We also considered the name `StorageSpan`, but that did not add much beyond the name `Span`. `Span` clearly identifies itself as a relative of C++'s `std::span`.

##### Adding `load` and `loadUnaligned` to `Span<UInt8>`on `Span<some BitwiseCopyable>` instead of adding `RawSpan`

> TKTKTK

Reviewer 8: Is this a TODO?

Author: Yes, it leaked out!

##### A more sophisticated approach to indexing

This is discussed more fully in the [indexing appendix](#Indexing) below.

## Future directions

##### Defining `BorrowingIterator` with support in `for` loops

> This proposal does not define an `IteratorProtocol` conformance, since it would need to be borrowed and non-escapable. This is not compatible with `IteratorProtocol`. As such, `Span`  is not directly usable in `for` loops as currently defined. A `BorrowingIterator` protocol for non-escapable and non-copyable containers must be defined, providing a `for` loop syntax where the element is borrowed through each iteration. Ultimately we should arrive at a way to iterate through borrowed elements from a borrowed view:

Reviewer 8: It's not clear to me that `Span` isn't already what a `BorrowingIterator` would be.

Reviewer 2: [Barring any last-minute changes] borrowing iteration is going to be based on API that incrementally exposes a _series_ of spans, not just a single one. So a `Span` can potentially be a borrowing iterator, but a borrowing iterator needs to be more than just one `Span` instance.

```swift
borrowing view: Span<Element> = ...
for borrowing element in view {
  doSomething(element)
}
```

In the meantime, it is possible to loop through a `Span`'s elements by direct indexing:

```swift
func doSomething(_ e: borrowing Element) { ... }
let view: Span<Element> = ...
// either:
var i = 0
while i < view.count {
  doSomething(view[i])
  view.index(after: &i)
}

// ...or:

> for i in 0..<view.indices {

Reviewer 2: ```suggestion
for i in 0 ..< view.count {
```

  doSomething(view[i])
}
```

##### Collection-like protocols for non-copyable and non-escapable types

> Non-copyable and non-escapable containers would benefit from a `Collection`-like protocol family to represent a set basic, common operations. This may be `Collection` if we find a way to make it work; it may be something else.

Reviewer 2: Let's not get people's hopes up. `Collection` is too deeply tied to copyability; we aren't working on its generalization (neither for noncopyable/nonescapable conforming types, nor noncopyable Elements).

Author: Removed the second sentence of this paragraph.

1. The pointer itself is unsafe and unmanaged
2. `subscript` is only bounds-checked in debug builds of client code
3. It might escape the duration of the closure
4. Exclusivity of writes is not enforced
5. Initialization of any particular memory address is not ensured

I.e., it is unsafe in all the ways `UnsafeBufferPointer`-passing closure APIs are unsafe in addition to being unsafe in exclusivity and in initialization.

Loading an uninitialized non-`BitwiseCopyable` value leads to undefined behavior. Loading an uninitialized `BitwiseCopyable` value does not immediately lead to undefined behavior, but it produces a garbage value which may lead to misbehavior of the program.

A `MutableSpan<T>` should provide a better, safer alternative to mutable memory in the same way that `Span<T>` provides a better, safer read-only type. `MutableSpan<T>` would also automatically enforce exclusivity of writes.

However, it alone does not track initialization state of each address, and that will continue to be the responsibility of the developer.

##### Delegating initialization of memory with `OutputSpan<T>`

> 

Reviewer 8: I'll update this with the current naming around `OutputSpan`.

Author: This section also needs to talk about the simple case where there is always an initialized prefix with an uninitialized tail. (will do)

Author: I've improved this section.

  return b.distance(from: b.startIndex, to: e)
}
let data = Data((0..<9).reversed()) // Data conforms to ContiguousStorage
let array = Array(data) // Array<UInt8> also conforms to ContiguousStorage
myStrnlen(data)  // 8
myStrnlen(array) // 8
```

This would probably consist of a new type of custom conversion in the language. A type author would provide a way to convert from their type to an owned `Span`, and the compiler would insert that conversion where needed. This would enhance readability and reduce boilerplate.

##### Interopability with C++'s `std::span` and with llvm's `-fbounds-safety`
The [`std::span`](https://en.cppreference.com/w/cpp/container/span) class template from the C++ standard library is a similar representation of a contiguous range of memory. LLVM may soon have a [bounds-checking mode](https://discourse.llvm.org/t/70854) for C. These are opportunities for better, safer interoperation with Swift, via a type such as `Span`.

## Acknowledgments

Joe Groff, John McCall, Tim Kientzle, Karoy Lorentey contributed to this proposal with their clarifying questions and discussions.

## <a name="Indexing"></a>Appendix: Index and slicing design considerations

> 

Reviewer 8: I'll need to rework this to read a little bit better.

> ## Safe Access to Contiguous Storage

Reviewer 9: ```suggestion
# Safe Access to Contiguous Storage
```
